dev/io.openliberty.io.smallrye.config/src/io/openliberty/microprofile/config/internal/extension/ConfigSourceWrapper.java

```diff
@@ -0,0 +1,156 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.microprofile.config.internal.extension;
+
+import java.util.AbstractMap;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.websphere.ras.annotation.Trivial;
+
+import io.openliberty.checkpoint.spi.CheckpointHook;
+import io.openliberty.checkpoint.spi.CheckpointPhase;
+
+public class ConfigSourceWrapper implements ConfigSource {
+    private static final TraceComponent tc = Tr.register(ConfigSourceWrapper.class);
+    private final ConfigSource source;
+    private final Map<String, String> propertiesAccessedOnCheckpoint;
+    private final CheckpointPhase phase;
+    private final AtomicBoolean hookAdded = new AtomicBoolean();
+
+    public ConfigSourceWrapper(ConfigSource source, CheckpointPhase phase) {
+        this.source = source;
+        propertiesAccessedOnCheckpoint = new HashMap<>();
```

Comment:
Should this be a thread-safe map?

I think it's unlikely that a config source will be accessed from multiple threads during startup but I'm not sure we can guarantee it.

Reply:
Added synchronization while doing read/write into the hashmap.

Reply:
Hmm, it looks right but if we're going to do it that way, can we add a javadoc comment to the field stating that reads and writes to the map must synchronize on the map?

Reply:
Added the javadoc comment. Thanks!